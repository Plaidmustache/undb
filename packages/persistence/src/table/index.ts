export * from "./table.query-repository"
export * from "./table.repository"
