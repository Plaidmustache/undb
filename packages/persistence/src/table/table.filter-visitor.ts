import { WontImplementException } from "@undb/domain"
import type {
  ITableSpecVisitor,
  TableDo,
  TableIdSpecification,
  TableNameSpecification,
  TableSchemaSpecification,
  TableViewsSpecification,
  WithNewFieldSpecification,
  WithUpdatedFieldSpecification,
  WithViewAggregate,
  WithViewColor,
  WithViewFilter,
  WithViewSort,
} from "@undb/table"
import type {
  TableFormsSpecification,
  WithFormSpecification,
  WithNewFormSpecification,
} from "@undb/table/src/specifications/table-forms.specification"
import type { WithTableRLS } from "@undb/table/src/specifications/table-rls.specification"
import { eq } from "drizzle-orm"
import { AbstractDBFilterVisitor } from "../abstract-db.visitor"
import { tables } from "../tables"

export class TableFilterVisitor extends AbstractDBFilterVisitor<TableDo> implements ITableSpecVisitor {
  withUpdatedField(spec: WithUpdatedFieldSpecification): void {
    throw new Error("Method not implemented.")
  }
  withForm(views: WithFormSpecification): void {
    throw new Error("Method not implemented.")
  }
  withNewField(schema: WithNewFieldSpecification): void {
    throw new Error("Method not implemented.")
  }
  withForms(views: TableFormsSpecification): void {
    throw new Error("Method not implemented.")
  }
  withNewForm(views: WithNewFormSpecification): void {
    throw new Error("Method not implemented.")
  }
  withViewAggregate(viewColor: WithViewAggregate): void {
    throw new Error("Method not implemented.")
  }
  withTableRLS(rls: WithTableRLS): void {
    throw new WontImplementException(TableFilterVisitor.name + ".withTableRLS")
  }
  withViewColor(viewColor: WithViewColor): void {
    throw new WontImplementException(TableFilterVisitor.name + ".withViewColor")
  }
  withViewSort(viewSort: WithViewSort): void {
    throw new WontImplementException(TableFilterVisitor.name + ".withViewSort")
  }
  withViewFilter(viewFilter: WithViewFilter): void {
    throw new WontImplementException(TableFilterVisitor.name + ".withViewFilter")
  }
  withViews(views: TableViewsSpecification): void {
    throw new WontImplementException(TableFilterVisitor.name + ".withViews")
  }
  withId(id: TableIdSpecification): void {
    this.addCond(eq(tables.id, id.id.value))
  }
  withName(name: TableNameSpecification): void {
    this.addCond(eq(tables.name, name.name.value))
  }
  withSchema(schema: TableSchemaSpecification): void {
    throw new WontImplementException(TableFilterVisitor.name + ".withSchema")
  }
}
