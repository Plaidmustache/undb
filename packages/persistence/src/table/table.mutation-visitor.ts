import { NotImplementException, WontImplementException } from "@undb/domain"
import type {
  ITableSpecVisitor,
  TableDo,
  TableIdSpecification,
  TableIdsSpecification,
  TableNameSpecification,
  TableSchemaSpecification,
  TableViewsSpecification,
  WithNewFieldSpecification,
  WithNewView,
  WithUpdatedFieldSpecification,
  WithView,
  WithViewAggregate,
  WithViewColor,
  WithViewFields,
  WithViewFilter,
  WithViewOption,
  WithViewSort,
} from "@undb/table"
import type {
  TableFormsSpecification,
  WithFormSpecification,
  WithNewFormSpecification,
} from "@undb/table/src/specifications/table-forms.specification"
import type { WithTableRLS } from "@undb/table/src/specifications/table-rls.specification"
import { AbstractDBMutationVisitor } from "../abstract-db.visitor"
import type { tables } from "../tables"

export class TableMutationVisitor
  extends AbstractDBMutationVisitor<TableDo, typeof tables>
  implements ITableSpecVisitor
{
  constructor(public readonly table: TableDo) {
    super()
  }
  withView(views: WithView): void {
    this.addUpdates({ views: this.table.views?.toJSON() })
  }
  withNewView(views: WithNewView): void {
    this.addUpdates({ views: this.table.views?.toJSON() })
  }
  withViewOption(viewOption: WithViewOption): void {
    this.addUpdates({ views: this.table.views?.toJSON() })
  }
  withViewFields(fields: WithViewFields): void {
    this.addUpdates({ views: this.table.views?.toJSON() })
  }
  withUpdatedField(spec: WithUpdatedFieldSpecification): void {
    this.addUpdates({ schema: this.table.schema?.toJSON() })
  }

  withForm(views: WithFormSpecification): void {
    this.addUpdates({ forms: this.table.forms?.toJSON() })
  }
  withForms(forms: TableFormsSpecification): void {
    this.addUpdates({ forms: this.table.forms?.toJSON() })
  }
  withNewForm(form: WithNewFormSpecification): void {
    this.addUpdates({ forms: this.table.forms?.toJSON() })
  }
  withNewField(schema: WithNewFieldSpecification): void {
    this.addUpdates({ schema: this.table.schema?.toJSON() })
  }
  withViewAggregate(viewColor: WithViewAggregate): void {
    this.addUpdates({ views: this.table.views?.toJSON() })
  }
  withTableRLS(rls: WithTableRLS): void {
    this.addUpdates({ rls: this.table.rls.into(undefined)?.toJSON() })
  }
  withViewSort(viewSort: WithViewSort): void {
    this.addUpdates({ views: this.table.views?.toJSON() })
  }
  withViewColor(viewFilter: WithViewColor): void {
    this.addUpdates({ views: this.table.views?.toJSON() })
  }
  withId(id: TableIdSpecification): void {
    throw new WontImplementException(TableMutationVisitor.name + ".withId")
  }
  withName(name: TableNameSpecification): void {
    throw new NotImplementException(TableMutationVisitor.name + ".withName")
  }
  withSchema(schema: TableSchemaSpecification): void {
    throw new NotImplementException(TableMutationVisitor.name + ".withSchema")
  }
  withViews(views: TableViewsSpecification): void {
    throw new NotImplementException(TableMutationVisitor.name + ".withSchema")
  }
  withViewFilter(viewFilter: WithViewFilter): void {
    this.addUpdates({ views: this.table.views?.toJSON() })
  }
  idsIn(ids: TableIdsSpecification): void {}
}
