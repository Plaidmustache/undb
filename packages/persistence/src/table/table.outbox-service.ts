import { injectContext, type IContext } from "@undb/context"
import { singleton } from "@undb/di"
import type { ITableOutboxService, TableDo } from "@undb/table"
import { getCurrentTransaction } from "../ctx"
import { OutboxMapper } from "../outbox.mapper"

@singleton()
export class TableOutboxService implements ITableOutboxService {
  constructor(
    @injectContext()
    private readonly context: IContext,
  ) {}
  async save(table: TableDo): Promise<void> {
    const trx = getCurrentTransaction()
    const values = table.domainEvents.map((e) => OutboxMapper.fromEvent(e, this.context))
    if (!values.length) return
    await trx.insertInto("undb_outbox").values(values).execute()
    table.removeEvents(table.domainEvents)
  }

  async saveMany(d: TableDo[]): Promise<void> {
    const trx = getCurrentTransaction()
    const values = d.flatMap((table) => table.domainEvents.map((e) => OutboxMapper.fromEvent(e, this.context)))
    if (!values.length) return
    await trx.insertInto("undb_outbox").values(values).execute()
    d.forEach((table) => table.removeEvents(table.domainEvents))
  }
}
