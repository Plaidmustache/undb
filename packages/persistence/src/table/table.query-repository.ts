import { singleton } from '@undb/di'
import { None, Option, Some } from '@undb/domain'
import {
  TableIdSpecification,
  type ITableDTO,
  type ITableQueryRepository,
  type TableComositeSpecification,
  type TableId,
} from '@undb/table'
import type { Database } from '../db'
import { injectDb } from '../db.provider'
import { tables } from '../tables'
import { TableDbQuerySpecHandler } from './table-db.query-spec-handler'
import type { TableMapper } from './table.mapper'
import { injectTableMapper } from './table.mapper.provider'

@singleton()
export class TableQueryRepository implements ITableQueryRepository {
  constructor(
    @injectDb()
    private readonly db: Database,
    @injectTableMapper()
    private readonly mapper: TableMapper
  ) {}

  async find(spec: Option<TableComositeSpecification>): Promise<ITableDTO[]> {
    const qb = this.db.select().from(tables).$dynamic()

    const result = await new TableDbQuerySpecHandler(qb).handle(spec)

    return result.map((r) => this.mapper.toDTO(r))
  }

  async findOneById(id: TableId): Promise<Option<ITableDTO>> {
    const qb = this.db.select().from(tables).$dynamic()

    const spec = Some(new TableIdSpecification(id))
    const tb = await new TableDbQuerySpecHandler(qb).handle(spec).limit(1)

    return tb.length ? Some(this.mapper.toDTO(tb[0])) : None
  }
}
