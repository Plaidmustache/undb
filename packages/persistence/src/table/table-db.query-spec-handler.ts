import { Option } from '@undb/domain'
import type { TableComositeSpecification } from '@undb/table'
import type { SQLiteSelectQueryBuilder } from 'drizzle-orm/sqlite-core'
import { TableQueryVisitor } from './table.query-visitor'

export class TableDbQuerySpecHandler<Q extends SQLiteSelectQueryBuilder> {
  constructor(private qb: Q) {}

  handle(spec: Option<TableComositeSpecification>): Q {
    if (spec.isSome()) {
      const visitor = new TableQueryVisitor()
      spec.unwrap().accept(visitor)

      this.qb = this.qb.where(visitor.cond)
    }

    return this.qb
  }
}
