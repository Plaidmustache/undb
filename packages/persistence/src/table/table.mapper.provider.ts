import { inject } from '@undb/di'
import { TableMapper } from './table.mapper'

export const injectTableMapper = () => inject(TableMapper)
