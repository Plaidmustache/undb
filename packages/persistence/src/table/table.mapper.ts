import { singleton } from '@undb/di'
import type { Mapper } from '@undb/domain'
import { injectTableBuilder, type ITableBuilder, type ITableDTO, type TableDo } from '@undb/table'
import type { NewTable, Table } from '../tables'

@singleton()
export class TableMapper implements Mapper<TableDo, NewTable, ITableDTO> {
  constructor(
    @injectTableBuilder()
    private readonly builder: ITableBuilder
  ) {}

  toDo(entity: Table): TableDo {
    return this.builder
      .setId(entity.id)
      .setName(entity.name)
      .setSchema(entity.schema ?? [])
      .build()
  }

  toEntity(domain: TableDo): NewTable {
    return {
      id: domain.id.value,
      name: domain.name.value,
      schema: domain.schema.toJSON(),
    }
  }

  toDTO(entity: Table): ITableDTO {
    return {
      id: entity.id,
      name: entity.name,
      schema: entity.schema ?? [],
    }
  }
}
