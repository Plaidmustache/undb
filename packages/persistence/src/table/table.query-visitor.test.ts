import { or } from '@undb/domain'
import { TableIdSpecification, TableIdVo, TableNameSpecification, TableNameVo } from '@undb/table'
import Database from 'bun:sqlite'
import { beforeEach, describe, expect, test } from 'bun:test'
import { drizzle } from 'drizzle-orm/bun-sqlite'
import { tables } from '../tables'
import { TableFilterVisitor } from './table.filter-visitor'

export const sqlite = new Database(':memory:')
const db = drizzle(sqlite)

describe('TableQueryVisitor', () => {
  let visitor: TableFilterVisitor

  beforeEach(() => {
    visitor = new TableFilterVisitor()
  })

  test.each([
    new TableIdSpecification(new TableIdVo('1')),
    new TableNameSpecification(new TableNameVo('table')),
    or(new TableIdSpecification(new TableIdVo('1')), new TableNameSpecification(new TableNameVo('table'))).unwrap(),
    new TableIdSpecification(new TableIdVo('1')).not(),
    or(
      new TableIdSpecification(new TableIdVo('1')).not(),
      new TableNameSpecification(new TableNameVo('table'))
    ).unwrap(),
  ])('should get correct query', (spec) => {
    spec.accept(visitor)

    const sql = db.select().from(tables).where(visitor.cond).toSQL()
    expect(sql).toMatchSnapshot()
  })
})
