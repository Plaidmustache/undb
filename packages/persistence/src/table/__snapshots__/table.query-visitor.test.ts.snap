// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`TableQueryVisitor should get correct query 1`] = `
{
  "params": [
    "tbl_1",
  ],
  "sql": "select "id", "name", "schema" from "undb_table" where "undb_table"."id" = ?",
}
`;

exports[`TableQueryVisitor should get correct query 2`] = `
{
  "params": [
    "table",
  ],
  "sql": "select "id", "name", "schema" from "undb_table" where "undb_table"."name" = ?",
}
`;

exports[`TableQueryVisitor should get correct query 3`] = `
{
  "params": [
    "tbl_1",
    "table",
  ],
  "sql": "select "id", "name", "schema" from "undb_table" where ("undb_table"."id" = ? or "undb_table"."name" = ?)",
}
`;

exports[`TableQueryVisitor should get correct query 4`] = `
{
  "params": [
    "tbl_1",
  ],
  "sql": "select "id", "name", "schema" from "undb_table" where not "undb_table"."id" = ?",
}
`;

exports[`TableQueryVisitor should get correct query 5`] = `
{
  "params": [
    "tbl_1",
    "table",
  ],
  "sql": "select "id", "name", "schema" from "undb_table" where (not "undb_table"."id" = ? or "undb_table"."name" = ?)",
}
`;
